features/telemetry/adapters/collection/src/main/java/org/opennms/netmgt/telemetry/adapters/collection/AbstractPersistingAdapter.java

```diff
@@ -126,17 +144,23 @@ private Package getPackageFor(Protocol protocol, CollectionAgent agent) {
                 // No filter specified, always match
                 return pkg;
             }
-            // TODO: This is really inefficient, since it actually retrieves *all*
-            // IP addresses that match the filter, and then checks of the given address
-            // is in the set. See HZN-1161.
-            // NOTE: The location of the host address is not taken into account.
-            if (filterDao.isValid(agent.getHostAddress(), filterRule)) {
+            if (isValid(agent.getHostAddress(), filterRule)) {
                 return pkg;
             }
         }
         return null;
     }
 
+    private boolean isValid(String hostAddress, String filterRule) {
+        try {
+            Optional<Boolean> validOptional = cache.get(new CacheKey(hostAddress, filterRule));
```

Comment:
Maybe we should cache the effective package used for the agent instead of the results of the filter match? This would save us from having to iterate over all of the packages and should reduce the cache size.

Reply:
Yeah, that makes sense. Also the CacheKey is missing a `equals()` and `hashCode()` implementation. Ooops.